### PR TITLE
Basic folder and data structures and pyvista plotting

### PR DESCRIPTION
**Pull request by banesullivan**
Just opening this PR to show the diff

In summary, there is a new `geometry` module that defines the spatial data structures of subsurface data. These data structures draw much inspiration from PyVista yet there is a fundamental difference we are implementing here: the idea of "meshes" and "grids".

There are two submodules for mesh and grid data structures. Mesh data structures will track all of their data/spatial references in Pandas DataFrame's while grid data structures will track all of their data/spatial reference in an xarray dataset.

Meshes are any subsurface data where all of the points of that data need to be explicitly defined. Examples of data that would be in the mesh form include: point clouds, line sets (wells), triangulated surfaces, structured grids (e.g. `discretize.CurviMesh` or `pyvista.StructuredGrid`), tetrahedralized volumes.

Grids are any subsurface data with implicitly defined points. Examples of these kinds of data include: rasters, rectilinear grids, (think seismic volumes), etc. The idea being here that the entire dataset can be defined by a few parameters like origin, U/V direction (orientation), and cell spacings. These kinds of data make sense to track in xarray.

Its important to note that I am classifying structured grids as meshes here because all of the points of those types of data structures must be known and cannot be implicitly defined.

### CODE REVIEW

**Comment by bluetyson** [comment]
So a reason to be a dataframe, not a geodataframe if tracking coordinates?

**Comment by banesullivan** [comment]
Geodataframes from geo pandas only supports 2D data AFIK. Is there 3D support?

**Comment by bluetyson** [comment]
Ah, yes, right!

File: subsurface/geometry/mesh.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+
+        Parameters
+        ----------
+        points : pd.DataFrame
+            A dataframe that has XYZ coordinates that are named as such.
+            Additional columns will be tracked as point data.
+
+        """
+        self._df_points = pd.DataFrame(columns=['X', 'Y', 'Z'])
+        self._df_point_data = pd.DataFrame()
```
Comment (Leguark) [inline subsurface/geometry/mesh.py:44]: is it worth it to have two dataframes or should we just make one and we just make views:

self._df = pd.Data...
self._df_points is self._df['X', 'Y', 'Z']
self._df_point data is self._df[the rest]

**Comment by florian-wagner** [comment]
Hi @banesullivan,

thanks for kicking of the data structure API by bringing in your valuable pv-experience. I agree that xarray's are predestined for structured grids and that a collection of pandas data frames (at least one for the nodes and one for the cells) are a good and flexible representation of general cell-based data sets.

With regard to the latter: Do we need different classes depending on the cell type? I agree that common functionality could be shared within something like `_CellDataMixin` to avoid redundancy, but is that necessary or could we directly aim for a more general `UnstructuredMesh` object (similar as in pyvista/VTK)?

pyGIMLi for example can handle various cell types (hex, tets, triangular prisms, quadratic pyramids, etc.) but more importantly a mixture of those (as the shape functions for FEM calculations are automatically generated based on the cell type). Are there reasons against a more general object where the number of columns in the cell data frame represents the maximum number nodes of a cell (let's say 6), and a tetrahedron would only use 4 of them with the unused two being NaNs?

Just thinking out loud here ;-)

Cheers
Florian

**Comment by Leguark** [comment]
I like that idea @florian-wagner. I have been thinking since Saturday how we can represent all geometries with the minimum number of - in this case - Dataframes (but as many columns as we need) and I reached to similar idea:
- Point data Dataframe [X, Y, Z] and properties
- Element data Dataframe [Int1, Int2, Int3, Int4, ..] and properties
    + Depending on the number of points per element (How many ints or whatever the official name is) we would have lines, triangles...
    + No matter the dimensionality the property will be located always in the center and will represent the whole volume

Am I missing something Florian?

**Comment by florian-wagner** [comment]
No, that's exactly what I had in mind. This would allow arbitrary geometries with point- and cell-based values in to compact pandas data frames. The first 8 columns in the cell dataframe would be reserved for node indices (not six as I wrote earlier...) and could be NaN if not used or columns Int5-Int8 could even be dropped when a mesh only containes tets for example.

When all libraries could read and write such a format + VTK I/O + pyvista viz, I would already be a happy user of `subsurface`.

**Comment by Leguark** [comment]
I think it is time to merge this. Some of the ideas that based this changes: https://github.com/softwareunderground/subsurface/blob/mig_dev/sdd.md

This is the summary of features:

- Created the full folder structure.
- Added base_structures (StructuredData and UnstructuredData)
- Adapted elements (points, lines, trisurf etc)
- Pyvista plotting for all the elements

File: requirements.txt [hunk 2]
```diff
@@ -1,6 +1,5 @@
...
-xarray
-segyio
-pytest
-matplotlib
-pyvista
+pandas~=1.1.1
+xarray~=0.16.0
+numpy~=1.19.1
+pytest~=6.0.1
+setuptools~=46.4.0
```
Comment (prisae) [inline requirements.txt:5]: Funny each requirement has one version that does not work. Why is that?

File: sdd.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
++ I/O 
+    + segyio
+    + welly
+    + rasterio
+    + ...
++ Visualization
+    + Matplotlib
+    + pyvista
++ Standard formats
+    + OMF
```
Comment (prisae) [inline sdd.md:27]: Adding geoh5py?

**Review by prisae — APPROVED** [review]
Just a few comments. I guess we will discuss more and Monday.

File: sdd.md (hunk 3 above)
Comment (Leguark) [reply sdd.md:27]: sure

File: requirements.txt (hunk 2 above)
Comment (Leguark) [reply requirements.txt:5]: I think those were the requirements in June. I did not know that meant "except", I will remove the versions thanks for the hint

**Review by prisae — APPROVED** [review]

**Pull request merged**